In [1]:
import pandas as pd

#Key Column: 
#Both DataFrames have FellowshipID and FirstName columns that can be used as keys.

#Merge Types:
#inner: Keeps only rows that exist in both DataFrames.
#outer: Keeps all rows from both DataFrames, filling missing values with NaN.
#left: Keeps all rows from the left DataFrame (df1) and matches data from the right DataFrame (df2).
#right: Keeps all rows from the right DataFrame (df2) and matches data from the left DataFrame (df1).

In [2]:
df1 =pd.read_csv('https://raw.githubusercontent.com/AlexTheAnalyst/PandasYouTubeSeries/refs/heads/main/LOTR.csv')
df1

FellowshipID FirstName     Skills
0          1001     Frodo     Hiding
1          1002   Samwise  Gardening
2          1003   Gandalf     Spells
3          1004    Pippin  Fireworks

In [3]:
df2 =   pd.read_csv('https://raw.githubusercontent.com/AlexTheAnalyst/PandasYouTubeSeries/refs/heads/main/LOTR%202.csv')
df2

FellowshipID FirstName   Age
0          1001     Frodo    50
1          1002   Samwise    39
2          1006   Legolas  2931
3          1007    Elrond  6520
4          1008  Barromir    51

## INNER MERGE
#Keeps only rows that exist in both DataFrames.

In [4]:
inner_merge = pd.merge(df1,df2, on='FellowshipID',how ='inner')
#so what this does is that, it merges the rows of both dataframe, containing same fellowship ids.
#since both datframe containes FirstName column, with matching data for the fellowship id 1001 and 1002 , it uses suffix _x and _y, to seperate 
#the first name column.

In [5]:
inner_merge

FellowshipID FirstName_x     Skills FirstName_y  Age
0          1001       Frodo     Hiding       Frodo   50
1          1002     Samwise  Gardening     Samwise   39

In [6]:
df1

FellowshipID FirstName     Skills
0          1001     Frodo     Hiding
1          1002   Samwise  Gardening
2          1003   Gandalf     Spells
3          1004    Pippin  Fireworks

In [7]:
df2

FellowshipID FirstName   Age
0          1001     Frodo    50
1          1002   Samwise    39
2          1006   Legolas  2931
3          1007    Elrond  6520
4          1008  Barromir    51

In [8]:
inner_merge =pd.merge(df1,df2, on ='FirstName',how ='inner')
#so what this does is that, it merges the rows of both dataframe, containing the same first names
#since both dataframe has column 'FellowshipId' with same ids, it uses suffix _x and _y  to seperate them


In [9]:
inner_merge

FellowshipID_x FirstName     Skills  FellowshipID_y  Age
0            1001     Frodo     Hiding            1001   50
1            1002   Samwise  Gardening            1002   39

In [10]:
#innermerge based on fellowshipid and firstname
inner_merge=  pd.merge(df1,df2, on=['FellowshipID','FirstName'],how ='inner')
#merges the rows of both dataframe, containing same fellowship ID and firstname,using innerjoin.


In [11]:
inner_merge

FellowshipID FirstName     Skills  Age
0          1001     Frodo     Hiding   50
1          1002   Samwise  Gardening   39

##  OuterMerge
#Keeps all rows from both DataFrames, filling missing values with NaN.

In [12]:
df1 =pd.read_csv('https://raw.githubusercontent.com/AlexTheAnalyst/PandasYouTubeSeries/refs/heads/main/LOTR.csv')
df1

FellowshipID FirstName     Skills
0          1001     Frodo     Hiding
1          1002   Samwise  Gardening
2          1003   Gandalf     Spells
3          1004    Pippin  Fireworks

In [13]:
df2 =   pd.read_csv('https://raw.githubusercontent.com/AlexTheAnalyst/PandasYouTubeSeries/refs/heads/main/LOTR%202.csv')
df2

FellowshipID FirstName   Age
0          1001     Frodo    50
1          1002   Samwise    39
2          1006   Legolas  2931
3          1007    Elrond  6520
4          1008  Barromir    51

In [14]:
outer_merge = pd.merge(df1, df2, on='FellowshipID', how='outer')
# Merges all rows from both DataFrames based on the 'FellowshipID' column.
#seperates the firstnames column using the suffixes
# Rows with no match in one of the DataFrames will have NaN values in the respective columns.

In [15]:
outer_merge


FellowshipID FirstName_x     Skills FirstName_y     Age
0          1001       Frodo     Hiding       Frodo    50.0
1          1002     Samwise  Gardening     Samwise    39.0
2          1003     Gandalf     Spells         NaN     NaN
3          1004      Pippin  Fireworks         NaN     NaN
4          1006         NaN        NaN     Legolas  2931.0
5          1007         NaN        NaN      Elrond  6520.0
6          1008         NaN        NaN    Barromir    51.0

In [27]:
outer_merge2 = pd.merge(df1,df2,on='FirstName',how='outer', sort =False)

| FirstName | FellowshipID_x | Skills     | FellowshipID_y | Age  |
|-----------|----------------|------------|----------------|------|
| Frodo     | 1001          | Hiding     | 1001          | 50   |
| Samwise   | 1002          | Gardening  | 1002          | 39   |
| Gandalf   | 1003          | Spells     | NaN           | NaN  |
| Pippin    | 1004          | Fireworks  | NaN           | NaN  |
| Legolas   | NaN           | NaN        | 1006          | 2931 |
| Elrond    | NaN           | NaN        | 1007          | 6520 |
| Boromir   | NaN           | NaN        | 1008          | 51   |

In [29]:
#the above table is the output for the outer_merg2, but pandas automatically sorts the names into alphabetical order, so we get  the
#output in the below way. same result,  diffrent format.
#for some reason sort=False is not working here??

In [28]:
outer_merge2

FellowshipID_x FirstName     Skills  FellowshipID_y     Age
0             NaN  Barromir        NaN          1008.0    51.0
1             NaN    Elrond        NaN          1007.0  6520.0
2          1001.0     Frodo     Hiding          1001.0    50.0
3          1003.0   Gandalf     Spells             NaN     NaN
4             NaN   Legolas        NaN          1006.0  2931.0
5          1004.0    Pippin  Fireworks             NaN     NaN
6          1002.0   Samwise  Gardening          1002.0    39.0

In [30]:
outer_merge3 =  pd.merge(df1,df2, on =['FellowshipID','FirstName'],how='outer', sort=False)

In [33]:
outer_merge3
#sort= False,  is working here..the result format  is correct

FellowshipID FirstName     Skills     Age
0          1001     Frodo     Hiding    50.0
1          1002   Samwise  Gardening    39.0
2          1003   Gandalf     Spells     NaN
3          1004    Pippin  Fireworks     NaN
4          1006   Legolas        NaN  2931.0
5          1007    Elrond        NaN  6520.0
6          1008  Barromir        NaN    51.0

## left join
#Keeps all rows from the left DataFrame (df1) and matches data from the right DataFrame (df2).

In [35]:
df1

FellowshipID FirstName     Skills
0          1001     Frodo     Hiding
1          1002   Samwise  Gardening
2          1003   Gandalf     Spells
3          1004    Pippin  Fireworks

In [36]:
df2

FellowshipID FirstName   Age
0          1001     Frodo    50
1          1002   Samwise    39
2          1006   Legolas  2931
3          1007    Elrond  6520
4          1008  Barromir    51

In [37]:
left_merge = pd.merge(df1,df2,on='FellowshipID',how= 'left')

In [38]:
left_merge

FellowshipID FirstName_x     Skills FirstName_y   Age
0          1001       Frodo     Hiding       Frodo  50.0
1          1002     Samwise  Gardening     Samwise  39.0
2          1003     Gandalf     Spells         NaN   NaN
3          1004      Pippin  Fireworks         NaN   NaN

In [43]:
left_merge = pd.merge(df1,df2,on='FirstName', how= 'left',sort=False)

In [44]:
left_merge

FellowshipID_x FirstName     Skills  FellowshipID_y   Age
0            1001     Frodo     Hiding          1001.0  50.0
1            1002   Samwise  Gardening          1002.0  39.0
2            1003   Gandalf     Spells             NaN   NaN
3            1004    Pippin  Fireworks             NaN   NaN

In [45]:
left_merge= pd.merge(df1,df2,on=['FellowshipID','FirstName'],how='left',sort= False)

In [46]:
left_merge

FellowshipID FirstName     Skills   Age
0          1001     Frodo     Hiding  50.0
1          1002   Samwise  Gardening  39.0
2          1003   Gandalf     Spells   NaN
3          1004    Pippin  Fireworks   NaN

In [47]:
## right join
#Keeps all rows from df2 and adds matching data from df1:

In [48]:
df1

FellowshipID FirstName     Skills
0          1001     Frodo     Hiding
1          1002   Samwise  Gardening
2          1003   Gandalf     Spells
3          1004    Pippin  Fireworks

In [49]:
df2

FellowshipID FirstName   Age
0          1001     Frodo    50
1          1002   Samwise    39
2          1006   Legolas  2931
3          1007    Elrond  6520
4          1008  Barromir    51

In [50]:
right_merge = pd.merge(df1,df2,on='FellowshipID',how='right')

In [51]:
right_merge

FellowshipID FirstName_x     Skills FirstName_y   Age
0          1001       Frodo     Hiding       Frodo    50
1          1002     Samwise  Gardening     Samwise    39
2          1006         NaN        NaN     Legolas  2931
3          1007         NaN        NaN      Elrond  6520
4          1008         NaN        NaN    Barromir    51

In [52]:
right_merge= pd.merge(df1,df2,on='FirstName',how='right',sort=False)

In [53]:
right_merge

FellowshipID_x FirstName     Skills  FellowshipID_y   Age
0          1001.0     Frodo     Hiding            1001    50
1          1002.0   Samwise  Gardening            1002    39
2             NaN   Legolas        NaN            1006  2931
3             NaN    Elrond        NaN            1007  6520
4             NaN  Barromir        NaN            1008    51

In [56]:
right_merge= pd.merge(df1,df2,on=['FellowshipID','FirstName'],how='right',sort=False)

In [57]:
right_merge

FellowshipID FirstName     Skills   Age
0          1001     Frodo     Hiding    50
1          1002   Samwise  Gardening    39
2          1006   Legolas        NaN  2931
3          1007    Elrond        NaN  6520
4          1008  Barromir        NaN    51